### PR TITLE
[MDETR] Refactor model classes

### PR DESCRIPTION
**Pull request by ebsmothers**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #200
* #199
* #198
* #181
* __->__ #180
* #179

Summary: Create new classes for phrase grounding and VQA tasks. Move contrastive alignment
projections out of the core model. Add new loss classes for VQA.

Test plan:
Note that we modified the test results in test_mdetr, but it is just a revert to the numbers
prior to PR #138 (where we initially added the contrastive alignment projections to the model class).
```
python -m pytest -vv
```
E2E test in subsequent PRs

Differential Revision: [D38160122](https://our.internmc.facebook.com/intern/diff/D38160122)

### CODE REVIEW

File: examples/mdetr/loss.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        norm_factor = mask.sum() if mask.any() else 1.0
+        losses[f"{k}_loss"] = (
+            F.cross_entropy(pred_dict[k], label_dict[k]).masked_fill(~mask, 0).sum()
+            / norm_factor
+        )
+
+    return losses
+
+
+class MDETRLoss(nn.Module):
```
Comment (RdoubleA) [inline examples/mdetr/loss.py:175]: you might've answered this already in a similar PR, but what's the reason these losses aren't in core? is it because it's only used for VQA/GQA?

File: examples/mdetr/model.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+    model_output: MDETRModelOutput
+    loss: Dict[str, Tensor]
+
+
+class MDETRForPhraseGrounding(nn.Module):
+    def __init__(
+        self,
+        model: MDETR,
+        contrastive_alignment_image_projection: nn.Module,
+        contrastive_alignment_text_projection: nn.Module,
```
Comment (RdoubleA) [inline examples/mdetr/model.py:249]: will the contrastive projections ever be anything other than nn.Linear?

File: examples/mdetr/utils/checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+):
+    mapped_state_dict = {}
+    for i in range(n_layers):
+        mapped_state_dict.update(map_layer(hf_state_dict, tt_state_dict, i))
+    return mapped_state_dict
+
+
+# The main function used to map from the MDETR state dict to the TorchMultimodal one
+# TODO: refactor to remove the explicit dependency on n_layers
+def map_mdetr_state_dict(
```
Comment (RdoubleA) [inline examples/mdetr/utils/checkpoint.py:96]: why not remap once and then save the checkpoint? then you don't need to include this code. or is this the code that's run once?

File: torchmultimodal/models/mdetr.py [hunk 4]
```diff
@@ -81,11 +79,10 @@ def __init__(
         pos_embed: nn.Module,
         text_projection: nn.Module,
         image_projection: nn.Module,
-        query_embed: nn.Module,
+        query_embed: nn.Embedding,
         bbox_embed: nn.Module,
         class_embed: nn.Module,
```
Comment (RdoubleA) [inline torchmultimodal/models/mdetr.py:84]: `bbox_embed` and `class_embed` are `Optional` in the docstrings but not here

File: torchmultimodal/models/mdetr.py [hunk 5]
```diff
@@ -55,9 +53,9 @@ class MDETR(nn.Module):
                 prior to the multimodal transformer.
             query_embed (nn.Module): Learned object query embeddings (used in
                 transformer decoder).
-            bbox_embed (nn.Module): Embedding mapping transformer outputs to
+            bbox_embed (Optional[nn.Module]): Embedding mapping transformer outputs to
                 bounding boxes.
-            class_embed (nn.Module): Embedding mapping transformer outputs to classes.
+            class_embed (Optional[nn.Module]): Embedding mapping transformer outputs to classes.
```
Comment (RdoubleA) [inline torchmultimodal/models/mdetr.py:58]: add `extra_query_embeddings` to docstring

File: torchmultimodal/models/mdetr.py [hunk 6]
```diff
@@ -140,37 +132,38 @@ def forward(self, images: List[Tensor], text: List[Tensor]) -> MDETROutput:
...
             self.image_projection(image_embeddings),
             image_mask,
             query_embed,
             pos,
             text_memory=text_memory_resized,
             text_attention_mask=text_attention_mask,
         )
-        final_hidden_state = transformer_outputs.decoder_hidden_states[-1]
+
+        if self.extra_query_embeddings is not None:
```
Comment (RdoubleA) [inline torchmultimodal/models/mdetr.py:150]: could you add comments explaining this block?

File: torchmultimodal/models/mdetr.py [hunk 7]
```diff
@@ -140,37 +132,38 @@ def forward(self, images: List[Tensor], text: List[Tensor]) -> MDETROutput:
...
+            extra_embeddings = transformer_output.decoder_hidden_states[
+                0, :, -n_extra_embeddings:
+            ]
+            transformer_output = transformer_output._replace(
```
Comment (RdoubleA) [inline torchmultimodal/models/mdetr.py:154]: this is confusing, could you maybe split this into two lines so it's more clear?

File: examples/mdetr/utils/checkpoint.py (hunk 3 above)
Comment (ebsmothers) [reply examples/mdetr/utils/checkpoint.py:96]: Oh yeah I removed it in the next PR, let me clean up the stack so it's taken out altogether

File: examples/mdetr/loss.py (hunk 1 above)
Comment (ebsmothers) [reply examples/mdetr/loss.py:175]: Because it includes the contrastive alignment loss, which relies on some stuff from transformers.

File: examples/mdetr/model.py (hunk 2 above)
Comment (ebsmothers) [reply examples/mdetr/model.py:249]: In practice probably not. I am just following the typing used in FLAVA, they also leave the projections as generic nn.Modules

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160122).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160122).

File: torchmultimodal/models/mdetr.py [hunk 8]
```diff
@@ -58,7 +56,8 @@ class MDETR(nn.Module):
             bbox_embed (nn.Module): Embedding mapping transformer outputs to
                 bounding boxes.
             class_embed (nn.Module): Embedding mapping transformer outputs to classes.
-
+            extra_query_embeddings (Optiona[nn.Embedding]): Additional query embeddings,
```
Comment (RdoubleA) [inline torchmultimodal/models/mdetr.py:59]: nit:
```suggestion
            extra_query_embeddings (Optional[nn.Embedding]): Additional query embeddings,
```

**Review by RdoubleA — APPROVED** [review]
just one minor typo but otherwise looks great

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160122).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160122).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160122).

File: torchmultimodal/models/mdetr.py [hunk 9]
```diff
@@ -19,19 +18,18 @@
...
 
 
-class MDETROutput(NamedTuple):
+class MDETRModelOutput(NamedTuple):
+    transformer_output: MDETRTransformerOutput
     pred_logits: torch.Tensor
     pred_boxes: torch.Tensor
-    projected_queries: torch.Tensor
-    projected_tokens: torch.Tensor
+    extra_embeddings: Optional[torch.Tensor]
```
Comment (ankitade) [inline torchmultimodal/models/mdetr.py:32]:
```suggestion
    extra_query_embeddings: Optional[torch.Tensor]
```

File: examples/mdetr/model.py [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+
+        target_boxes = [t["boxes"] for t in targets]
+        indices = self.matcher(
+            model_output.pred_logits,
+            model_output.pred_boxes,
+            target_boxes,
+            positive_map,
+        )
+
+        loss = self.loss(
```
Comment (ankitade) [inline examples/mdetr/model.py:305]: what happens in eval mode though?

File: examples/mdetr/loss.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+                loss_dict.update(vqa_loss(vqa_preds, vqa_labels, vqa_masks))
+
+        if weight_dict is not None:
+            total_loss = self.total_losses_with_weights(loss_dict, weight_dict)
+            loss_dict.update(total_loss=total_loss)
+
+        return loss_dict
+
+
+def build_weight_dict(
```
Comment (ankitade) [inline examples/mdetr/loss.py:285]: used where?

File: examples/mdetr/model.py [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+    masked_dict_cross_entropy,
+    MDETRLoss,
+)
+from examples.mdetr.matcher import HungarianMatcher
+from torch import nn, Tensor
+from torchmultimodal.models.mdetr import MDETR, mdetr_resnet101, MDETRModelOutput
+from torchmultimodal.modules.losses.mdetr import box_losses, soft_token_prediction_loss
+
+
+class MultiHead(nn.Module):
```
Comment (ankitade) [inline examples/mdetr/model.py:24]: do u really need this? isnt moduledict enough?

File: examples/mdetr/model.py [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+        contrastive_alignment_text_projection,
+    )
+
+
+class MDETRPhraseGroundingOutput(NamedTuple):
+    model_output: MDETRModelOutput
+    loss: Dict[str, Tensor]
+
+
+class MDETRForPhraseGrounding(nn.Module):
```
Comment (ankitade) [inline examples/mdetr/model.py:252]: generally should considering splitting out into smaller diffs for different tasks

File: examples/mdetr/model.py (hunk 10 above)
Comment (ebsmothers) [reply examples/mdetr/model.py:305]: Will separate this out

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 14]
```diff
@@ -142,12 +142,16 @@ def forward(
...
     norm_layer: Callable[..., nn.Module] = FrozenBatchNorm2d,
+    freeze_weights: bool = True,
 ) -> MaskedIntermediateLayer:
     body = resnet101(
         replace_stride_with_dilation=[False, False, False],
         weights=weights,
         norm_layer=norm_layer,
     )
-
+    if freeze_weights:
```
[the remark this answers is not in the document]
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:152]: Like in the training loop? That feels awkward to me

File: examples/mdetr/loss.py (hunk 11 above)
Comment (ebsmothers) [reply examples/mdetr/loss.py:285]: Used in a subsequent PR

File: examples/mdetr/model.py (hunk 12 above)
Comment (ebsmothers) [reply examples/mdetr/model.py:24]: Yeah maybe, lemme see about replacing this in a follow-up
